components/director/internal/domain/scenarioassignment/engine.go

```diff
@@ -50,21 +66,86 @@ func (e *engine) EnsureScenarioAssigned(ctx context.Context, in model.AutomaticS
 	}
-	labels, err := e.labelRepo.GetScenarioLabelsForRuntimes(ctx, in.Tenant, runtimesIDs)
+
+	runtimeLabels, err := e.labelRepo.GetScenarioLabelsForRuntimes(ctx, in.Tenant, runtimesIDs)
 	if err != nil {
 		return errors.Wrap(err, "while fetching scenarios labels for matched runtimes")
 	}
 
-	labels = e.appendMissingScenarioLabelsForRuntimes(in.Tenant, runtimesIDs, labels)
-	return e.upsertScenarios(ctx, in.Tenant, labels, in.ScenarioName, e.uniqueScenarios)
+	err = e.addNewScenarioToExistingBundleInstanceAuthFromMatchedApplication(ctx, runtimeLabels, in.ScenarioName)
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("while adding new scenario to already existing bundle instance auth between app and runtime matched by scenario: "))
+	}
+
+	runtimeLabels = e.appendMissingScenarioLabelsForRuntimes(in.Tenant, runtimesIDs, runtimeLabels)
+	for _, label := range runtimeLabels {
+		rmtLabelInput, err := labelpkg.MergeScenarios(label, []string{in.ScenarioName}, labelpkg.UniqueScenarios)
+		if err != nil {
+			return err
+		}
+
+		if rmtLabelInput == nil {
+			return errors.New("unable to update scenarios for runtime label")
+		}
+
+		err = e.labelService.UpsertLabel(ctx, in.Tenant, rmtLabelInput)
+		if err != nil {
+			return errors.Wrapf(err, "while upserting scenarios label for runtime with id: %s", label.ObjectID)
+		}
+	}
+	return nil
+}
+
+func (e *engine) addNewScenarioToExistingBundleInstanceAuthFromMatchedApplication(ctx context.Context, runtimeLabels []model.Label, scenario string) error {
+	for _, runtimeLabel := range runtimeLabels {
+		runtimeScenarios, err := labelpkg.GetScenariosAsStringSlice(runtimeLabel)
+		if err != nil {
+			return errors.Wrap(err, "while parsing runtime label value")
+		}
+
+		inputScenarios := []string{scenario}
+		inputScenarios = append(inputScenarios, runtimeScenarios...)
+		if err = e.bundleInstanceAuthService.AssociateBundleInstanceAuthForNewRuntimeScenarios(ctx, runtimeScenarios, inputScenarios, runtimeLabel.ObjectID); err != nil {
+			return err
+		}
+	}
+
+	return nil
 }
 
 func (e *engine) RemoveAssignedScenario(ctx context.Context, in model.AutomaticScenarioAssignment) error {
 	labels, err := e.labelRepo.GetRuntimeScenariosWhereLabelsMatchSelector(ctx, in.Tenant, in.Selector.Key, in.Selector.Value)
 	if err != nil {
 		return errors.Wrap(err, "while getting runtimes scenarios which match given selector")
 	}
-	return e.upsertScenarios(ctx, in.Tenant, labels, in.ScenarioName, e.removeScenario)
+	for _, label := range labels {
+		runtimeID := label.ObjectID
+
+		if err = e.validateNoBundleInstanceAuthsExist(ctx, in.Tenant, runtimeID, in.ScenarioName); err != nil {
+			return err
+		}
+	}
+
+	for _, label := range labels {
+		rmtLabelInput, err := labelpkg.MergeScenarios(label, []string{in.ScenarioName}, e.removeScenario)
+		if err != nil {
+			return err
+		}
+
+		if rmtLabelInput == nil {
+			err := e.labelRepo.Delete(ctx, in.Tenant, label.ObjectType, label.ObjectID, model.ScenariosKey)
```

Comment:
Since the method works with scenarios it is more visible that we try to delete scenario label. Also, if (for any reason) there is a bug somewhere in above backend calls it will ensure that we won't damage our data.